File: RetrofitSample/app/src/main/java/com/xyz/networkcalls/MainActivity.java
package com.xyz.networkcalls;

import android.os.Bundle;
import android.widget.Toast;

import androidx.annotation.NonNull;
import androidx.appcompat.app.AppCompatActivity;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;

import java.util.ArrayList;
import java.util.List;

import retrofit2.Call;
import retrofit2.Callback;

public class MainActivity extends AppCompatActivity {

    private GeneralAdapter generalAdapter;
    private List<PhotosResponse> responseList = new ArrayList<>();
    private RecyclerView recyclerView;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);
        recyclerView = findViewById(R.id.recyclerView);
        makeApiCall();
        setRecyclerAdapter();
    }

    private void makeApiCall() {
        /*
        Initialize the API service calls which has all the Api calls
         */
        ApiService apiService = Network.getRetrofitInstance().create(ApiService.class);

        /*
        Make an API call and save the response in call object
         */
        Call<List<PhotosResponse>> call = apiService.getPhotos();

        /*
         * This enqueue method will make an asynchronous call and fetches the response from the server
         */
        call.enqueue(new Callback<List<PhotosResponse>>() {
            @Override
            public void onResponse(@NonNull Call<List<PhotosResponse>> call, @NonNull retrofit2.Response<List<PhotosResponse>> response) {
                /*
                This method is invoked if the response is successful
                 */
                if (response.code() == 200) {
                    responseList = response.body();
                    /*
                    Since this Api call is made in the background thread, there is a possibility that the response comes late,
                    so once the response comes we need to update the list and notify the adapter about the changes
                     */
                    generalAdapter.updateData(responseList);
                }
            }

            @Override
            public void onFailure(@NonNull Call<List<PhotosResponse>> call, @NonNull Throwable t) {
                /*
                If the Api call is failed then an error is thrown and it can be obtained from the throwable object
                 */
                Toast.makeText(MainActivity.this, "Failed " + t.getMessage(), Toast.LENGTH_SHORT).show();
            }
        });

    }

    private void setRecyclerAdapter() {
        generalAdapter = new GeneralAdapter(responseList);
        LinearLayoutManager layoutManager = new LinearLayoutManager(this);
        recyclerView.setLayoutManager(layoutManager);
        recyclerView.setAdapter(generalAdapter);
    }
}

File: evaluations/evaluation_week16/app/src/main/java/com/xyz/evaluations_week16/MainActivity.kt
package com.xyz.evaluations_week16

import android.os.Bundle
import androidx.appcompat.app.AppCompatActivity
import androidx.recyclerview.widget.LinearLayoutManager
import kotlinx.android.synthetic.main.activity_main.*
import retrofit2.Call
import retrofit2.Callback
import retrofit2.Response

class MainActivity : AppCompatActivity() {

    private var dataModelList = listOf<DataModel>()

    private lateinit var resourceAdapter: ResourceAdapter

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)
        setRecyclerAdapter()
        callAPI()
    }

    private fun setRecyclerAdapter() {
        val linearLayoutManager = LinearLayoutManager(this)
        resourceAdapter = ResourceAdapter(dataModelList)
        recyclerView.apply {
            layoutManager = linearLayoutManager
            adapter = resourceAdapter
        }
    }

    private fun callAPI() {
        val apiClient = Network.getInstance().create(ApiClient::class.java)
        val call = apiClient.getResources()
        call.enqueue(object : Callback<ResponseModel>{
            override fun onResponse(call: Call<ResponseModel>, response: Response<ResponseModel>) {
                response.body()?.let { it ->
                    //Since the response is nullable here use null safety operator so it is no null
                    it.data?.let {
                        /*
                        since the it.data is List<DataModel?>? use a null safety operator and make it
                         non nullable i.e List<DataModel?>
                         */
                        dataModelList = it as List<DataModel> // casting is required here as dataModelList is of type List<DataModel> where as we are receiving the type List<DataModel?>
                        resourceAdapter.updateData(dataModelList)
                    }
                }
            }

            override fun onFailure(call: Call<ResponseModel>, t: Throwable) {
                //Handle the response failure here
            }

        })

    }
}